Subarray introduction
total no. of subarrays = n +(n-1) + (n-2) + ..+3+2+1 = n*(n+1)/2

In [75]:
A = [1,2,3,4,5]
class SolutionSubarrayIntro:
    def __init__(self,array_p):
        self.A = array_p

    def subarray_print(self,start,end):
        ans =[]
        for i in range(start,end+1):
            ans.append(self.A[i])
        return ans
    
    def subarray_all_print(self):
        n = len(self.A)
        ans =[]
        for i in range(n): # start   
            for j in range(i,n): # end 
                row=[]
                for k in range(i,j+1):
                    row.append(A[k])
                ans.append(row)
        return ans
    
    def subarray_all_print_newline(self):
        n = len(self.A)
        for i in range(n):
            for j in range(i,n):
                #row=[]
                for k in range(i,j+1):
                    #row.append(A[k])
                    print(A[k],end=" ")
                print()
    def subarray_sum_bruteforce(self):
        n = len(self.A)
        
        for i in range(n):# start index 
            for j in range(i,n): # end index
                sum_s=0
                for k in range(i,j+1):
                    sum_s+= A[k]
                print(sum_s)

    def subarray_sum_prefixsum(self):
        def prefix_sum():
            n = len(self.A)
            sum_p = 0
            P = []
            for i in range(n):
                sum_p += self.A[i]
                P.append(sum_p)
            return P
        P = prefix_sum()
        n = len(self.A)
        for i in range(n):
            for j in range(i,n):
                if i == 0:
                    print(P[j])
                else:
                    print(P[j]-P[i-1])

    def subarray_sum_carryforward(self,start,end):
        #n = len(self.A)
        sum_c = 0
        for i in range(start,end+1):
            sum_c += self.A[i]
            print(sum_c)

    
    def subarray_all_sum_carryforward(self):
        n = len(self.A)
        for i in range(n):# i = start index 
            self.subarray_sum_carryforward(i,n-1)

    def subarray_all_sum_carryforward1(self):
        n = len(self.A)
        ans =[]
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans.append(sum_c)
                #print(sum_c)
        return ans
    
    def subarray_maxsum_carryforward(self):
        n = len(self.A)
        ans = float('-inf')
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans = max(ans,sum_c)

        return ans
    
    def subarray_minsum_carryforward(self):
        n = len(self.A)
        ans = float('+inf')
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans = min(ans,sum_c)

        return ans
    
    def subarray_evensum_carryforward(self):
        n = len(self.A)
        count = 0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                if (sum_c % 2 == 0):
                    count += 1
                #print(sum_c)
        return count
    
    def subarray_oddsum_carryforward(self):
        n = len(self.A)
        count = 0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                if (sum_c % 2 != 0):
                    count += 1
                #print(sum_c)
        return count
    
    def subarray_evenlengthsum_gt_B_carryforward(self,B):
        n = len(self.A)
        count = 0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                length = j-i + 1
                if (length % 2 == 0 and sum_c > B):
                    count += 1
                #print(sum_c)
        return count
    
    def subarray_all_globalsum_carryforward(self):
        n = len(self.A)
        ans =[]
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                ans.append(sum_c)
                #print(sum_c)
        return sum(ans)
    
    def subarray_all_globalsum_carryforward1(self):
        n = len(self.A)
        gsum=0
        for i in range(n):
            sum_c = 0
            for j in range(i,n):
                sum_c += self.A[j]
                gsum += sum_c
                #print(sum_c)
        return gsum
    
    def subarray_all_globalsum_prefixsum(self):
        def prefix_sum(array):
            n = len(array)
            sum_p = 0
            P = []
            for i in range(n):
                sum_p += array[i]
                P.append(sum_p)
            return P
        P = prefix_sum(self.A)
        n = len(self.A)
        P1 =[]
        for i in range(n):
            for j in range(i,n):
                if i == 0:
                    P1.append(P[j])
                else:
                    P1.append(P[j]-P[i-1])
        P2 = prefix_sum(P1)
        return P2[-1]
    
    def subarray_globalsum_optimized(self):
        gsum = 0
        n = len(self.A)
        for i in range(n):
            gsum += self.A[i]*(n-i)*(i+1)
        return gsum
    
    def subarray_index_occurrence(self,index):
        n = len(self.A)
        return (n-index)*(index+1)









    
s1 = SolutionSubarrayIntro(A)
#s1.subarray_print(2,4)
#s1.subarray_all_print()
s1.subarray_all_print_newline()
#s1.subarray_sum_bruteforce()
#print("Check")
#s1.subarray_sum_prefixsum()
#s1.subarray_sum_carryforward(2,4)
#s1.subarray_all_sum_carryforward()
#s1.subarray_all_sum_carryforward1()
#s1.subarray_maxsum_carryforward()
#s1.subarray_minsum_carryforward()
#s1.subarray_evensum_carryforward()
#s1.subarray_oddsum_carryforward()
#s1.subarray_evenlengthsum_gt_B_carryforward(3)
#s1.subarray_all_globalsum_carryforward()
#s1.subarray_all_globalsum_prefixsum()
#s1.subarray_all_globalsum_carryforward1()
s1.subarray_globalsum_optimized()
s1.subarray_index_occurrence(3)


1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 
2 
2 3 
2 3 4 
2 3 4 5 
3 
3 4 
3 4 5 
4 
4 5 
5 


8

In [46]:
A=[1,2,3,4,5]
n=len(A)
c=[]
for i in range(n):
    c.append((n-i)*(i+1))
sum_a = 0
for i in range(len(A)):
    sum_a += A[i]*(n-i)*(i+1)
print(sum_a)


105


total no. of Subarrays of len k = n-k+1
Total no. of occurrence of any index i (0 based ) in an array of length n with total n*(n+1)/2 = (n-i)*(i+1)

In [171]:
A = [1,2,3,4,5]
A =[-7,3,4,-2,5,3,-20,8,2,-10,4]
class SolutionSubarrayAdv:
    def __init__(self,array_p):
        self.A = array_p

    def subarray_maxsum_len(self,k):#sliding window
        n = len(self.A)
        ans=float('-inf')
        sum_k = 0
        for i in range(k):
            sum_k += self.A[i]
        ans = max(ans,sum_k)
        for j in range(1,n-k+1):
            sum_k = sum_k - self.A[j-1] + self.A[j+k-1]
            ans= max(ans,sum_k)
        return ans
    
    def subarray_maxsum_len_k_1(self,k):
        """
        Q. Find the maxsubarray sum of given length k for array A of n elements.
        
        Concepts :: 
        Total no. of subarrays = n*(n+1)/2
        Total no. of subarrays of length k = n-k+1
        first subarray = 0 to k-1 . <=> k-1 -0 +1 = k
        last subarray = n-k to n-1. <=> n-1 -x +1 = k => x = n-k

        Optimization :
        Sliding Window Algo
        1. run loop for first k elements and save the sum as max.
        2. Slide window to the right by adding new element from the end and remove the element from the beginning . 
        end - start + 1 = k => end = start + k -1
        take the max of two - one is saved max and second the new sum.  
        """
        n = len(self.A) # length of the Array
        sum_a = 0
        for i in range(k): # o to k-1 of length k. first subarray. i denotes the index here from 0 to k-1
            sum_a += self.A[i]
        ans = sum_a # this is the initial max from first subarray from 0 to k-1 of length k
        for start in range(1,n-k+1):# last subarray is n-k to n-1. so this loop runs from 1 to n-k+1.then it will run from 1 to n-k. 0th index already taken.
            end = start +k -1 # end -start + 1 = k . This is new end.
            sum_a += self.A[end] - self.A[start-1] # adding the new end. removing the last start.
            ans = max(ans,sum_a) # taking the max of last max and new sum. 
        return ans # returning the final ans as final max.
    
    def subarray_maxsum_bruteforce(self):
        n = len(self.A)
        ans = float('-inf')
        for start in range(n):
            for end in range(start,n):
                sum_m = 0
                for i in range(start,end+1):
                    sum_m+=self.A[i]
                if sum_m > ans:
                    ans = sum_m
        return ans
    
    def subarray_maxsum_kadane_algo(self):
        sum_k = 0
        ans = float('-inf') # mas sum initiation
        for i,x in enumerate(self.A):
            sum_k += x
            # the below line can be written as : ans = max(sum_k,ans)
            if sum_k > ans:
                ans = sum_k
            # the below line can be writtne as : sum_k = max(sum_k,0)
            if sum_k < 0:
                sum_k =0
        return ans
    
    def subarray_maxsum_kadane_algo_with_indices(self):
        sum_k = 0
        ans = float('-inf')
        final = []
        best_start= 0
        best_end = 0
        flagS = 0
        flagP = 0
        flagN = 0
        state =[]
        for i,x in enumerate(self.A):
            sum_k+=x
            if sum_k < ans:
                flagS=1
                final.append(f"sustainTrend {i} {ans} {sum_k}")
            if sum_k > ans:
                ans = sum_k
                if i == 0:
                    best_start = i
                    flagP = 2
                else:
                    best_end = i
                    flagP=2

                final.append(f"posTrend {i} {ans} {sum_k}")
            if sum_k <0 :
                sum_k = 0
                flagN = 3
                
                final.append(f"negTrend {i} {ans} {sum_k}")

            
            state.append([flagS,flagP,flagN])
            #print(f"possible {best_start} {best_end}")
        return final,state,ans

    # pos > neg > pos then take latest pos.
    # pos > neg > sus > neg then take the pos
    # pos > neg > sus > neg > pos then take the latest pos
    # pos > neg then take the pos
    # pos > neg > sus then take the pos
    def subarray_alternative_print_all(self): # even indices subarray . odd indices subarray 
        pass

    def subarray_fromright_print_all(self):
        pass

    def subarray_zero_queries(self,n,q):
        F = [0]*n
        #print(F)
        for start,value in q:
            F[start]+=value
        #print(F)
        for i in range(1,n):
            F[i]+=F[i-1]
        return F
    
    def subarray_zero_queries_with_end(self,n,q):
        ans = [0]*n
        print(ans)
        for start,end,value in q:
            ans[start]+=value
            if end <= n-2:
                ans[end+1]-=value
        print(ans)
        for i in range(1,n):
            ans[i]+=ans[i-1]
        print(ans)


"""
A =[-7,3,4,-2,5,3,-20,8,2,-10,4]
A=[-7,8]
A=[-7,-2]
A=[-7,-9]
A=[-7]
A=[-7,-2,-2,-1,-4,-3]
A =[-7,-8,3,4,-2,5,3,-20,8,2,-10,4]
A = [2,3,5,8]
A=[-4,-6,-2,-7,-3,-10]

-4
1. -2, +5
> 2 , 8
1.1. -1, 3 or 0
1.2. -6 
2. -7
"""
Q = [[1,3],[0,4],[2,9],[0,7],[4,-6],[8,-3]]
Q1 = [[1,4,10],[0,2,20],[7,9,40],[3,8,30],[0,5,50]]
#A=[-7,-8,-3]
s2 = SolutionSubarrayAdv(A)
s2.subarray_maxsum_len(3)
#s2.subarray_maxsum_bruteforce()
#s2.subarray_maxsum_kadane_algo()
#s2.subarray_maxsum_kadane_algo_with_indices()
#s2.subarray_zero_queries(10,Q)
#s2.subarray_zero_queries_with_end(10,Q1)
#s2.subarray_maxsum_len_k_1(3)

7

In [77]:
A =[3,4,-2,5,3,-20,8,2,-10,4]
for i,x in enumerate(A):
    print(i,x)

0 3
1 4
2 -2
3 5
4 3
5 -20
6 8
7 2
8 -10
9 4


Q.3. Prefix maximum
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
Kamal is a software developer and he's working on a new feature for an e-commerce website. The website has a list of prices for different products, and Kamal needs to find the maximum price of all products up to a given index.

He has the list of prices in an array A of length N, and he needs to write a program that will return the maximum price occurring in the subarray from 0 to i for every index i. Kamal needs your help to implement this function.



Problem Constraints
1 <= N <= 105
1 <= A[i] <= 109


Input Format
Only argument A is an array of integers.


Output Format
Return an array of integers where the i-th element of the array represents the maximum value occurring in the subarray from 0 to i in the input array A.


Example Input
Input 1:

A = [9, 7, 6, 18, 2]
Input 2:

A = [16, 8, 24, 9, 25, 17]


Example Output
Output 1:

[9, 9, 9, 18, 18]
Output 2:

[16, 16, 24, 24, 25, 25]


Example Explanation
Explanation 1:

The prefix maximum array of [9, 7, 6, 18, 2] is [9, 9, 9, 18, 18].
Explanation 2:

The prefix maximum array of [16, 8, 24, 9, 25, 17] is [16, 16, 24, 24, 25, 25].

In [123]:
class SolutionPrefixMax:
    def __init__(self,array_p):
        self.A = array_p

    def prefix_max(self):
        n = len(self.A)
        P = []
        P.append(self.A[0])
        for i in range(1,n):
            P.append(max(P[i-1],self.A[i]))
        return P
    
    def prefix_max_modifying_array(self):
        n = len(self.A)
        for i in range(1,n):
            self.A[i] = max(self.A[i],self.A[i-1])
        return self.A
    
A = [9, 7, 6, 18, 2]
s3 = SolutionPrefixMax(A)
#s3.prefix_max()
s3.prefix_max_modifying_array()

[9, 9, 9, 18, 18]

Q.4. Suffix maximum
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
Given an array A of length N, For every index i, you need to find the maximum value in subarray from i to N-1.


Problem Constraints
1 <= N <= 105
1 <= A[i] <= 109


Input Format
Only argument A is an array of integers.


Output Format
Return an array of integers denoting the suffix maximum of the given array.


Example Input
Input 1:

A = [12, 5, 6, 7]
Input 2:

A = [15, 9, 7, 11, 10]


Example Output
Output 1:

[12, 7, 7, 7]
Output 2:

[15, 11, 11, 11, 10]


Example Explanation
Explanation 1:

The suffix maximum array of [12, 5, 6, 7]] is [12, 7, 7, 7].
Explanation 2:

The suffix maximum array of [15, 9, 7, 11, 10] is [15, 11, 11, 11, 10].

In [60]:
class SolutionSuffixMax:
    def __init__(self,array_p):
        self.A = array_p

    def suffix_max(self):
        n = len(self.A)
        S=[]
        S.append(self.A[n-1])
        for i in reversed(range(n-1)):
            S.append(max(S[n-2-i],self.A[i]))
        return S[::-1]

A = [12, 5, 6, 7]       
s4 = SolutionSuffixMax(A)
s4.suffix_max()

[12, 7, 7, 7]

Q.5. Continuous Sum Query
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
There are A beggars sitting in a row outside a temple. Each beggar initially has an empty pot. When the devotees come to the temple, they donate some amount of coins to these beggars. Each devotee gives a fixed amount of coin(according to their faith and ability) to some K beggars sitting next to each other.

Given the amount P donated by each devotee to the beggars ranging from L to R index, where 1 <= L <= R <= A, find out the final amount of money in each beggar's pot at the end of the day, provided they don't fill their pots by any other means.
For ith devotee B[i][0] = L, B[i][1] = R, B[i][2] = P, Given by the 2D array B


Problem Constraints
1 <= A <= 2 * 105
1 <= L <= R <= A
1 <= P <= 103
0 <= len(B) <= 105


Input Format
The first argument is a single integer A.
The second argument is a 2D integer array B.


Output Format
Return an array(0 based indexing) that stores the total number of coins in each beggars pot.


Example Input
Input 1:-
A = 5
B = [[1, 2, 10], [2, 3, 20], [2, 5, 25]]


Example Output
Output 1:-
10 55 45 25 25


Example Explanation
Explanation 1:-
First devotee donated 10 coins to beggars ranging from 1 to 2. Final amount in each beggars pot after first devotee: [10, 10, 0, 0, 0]
Second devotee donated 20 coins to beggars ranging from 2 to 3. Final amount in each beggars pot after second devotee: [10, 30, 20, 0, 0]
Third devotee donated 25 coins to beggars ranging from 2 to 5. Final amount in each beggars pot after third devotee: [10, 55, 45, 25, 25]

In [121]:
A = 5
B = [[1, 2, 10], [2, 3, 20], [2, 5, 25]]
def subarray_zero_queries_with_end(n,q):# 1-based indexing.
        ans = [0]*n
        print(ans)
        for start,end,value in q:
            ans[start-1]+=value
            if end <= n-1:
                ans[end]-=value
        print(ans)
        for i in range(1,n):
            ans[i]+=ans[i-1]
        print(ans)
subarray_zero_queries_with_end(A,B)
    



[0, 0, 0, 0, 0]
[10, 45, -10, -20, 0]
[10, 55, 45, 25, 25]


Range Sum - II
Solved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
You are given an integer array A of size N. You need to perform Q queries on the given array and return the final array after processing all the queries.

Each query is of the form (l, r, c), where l and r are indices (1-based) representing a range in the array A, and c is an integer value.

For each query, you are required to add the value c to every element within the range [l, r] (inclusive).



Problem Constraints
1 <= N, Q <= 105

-106 <= A[i], c <= 106

1 <= l <= r <= N



Input Format
First argument is an integer array A of size N.

Second argument is a 2d array B of size Q x 3, where B[i][0] = l, B[i][1] = r and B[i][2] = c.



Output Format
Return the final array after processing all queries.



Example Input
Input 1:

 A = [1, 2, 1, 4]
 B = [
       [2, 3, 2]
       [1, 4, 5]
       [4, 4, 1]
     ]
Input 2:

 A = [1, 1, 2]
 B = [
       [2, 3, 2]
       [1, 3, 5]
     ]


Example Output
Output 1:

 [6, 9, 8, 10]
Output 2:

 [6, 8, 9]


Example Explanation
Explanation 1:

 First query, l = 2, r = 3, c = 2, after processing array becomes : [1, 4, 3, 4]
 Second query, l = 1, r = 4, c = 5, after processing array becomes : [6, 9, 8, 9]
 Third query, l = 4, r = 4, c = 1, after processing array becomes : [6, 9, 8, 10]
 Final array is [6, 9, 8, 10]
Explanation 2:

 First query, l = 2, r = 3, c = 2, after processing array becomes : [1, 3, 4]
 Second query, l = 1, r = 3, c = 5, after processing array becomes : [6, 8, 9]
 Final array is [6, 8, 9]

In [134]:
A = [1, 2, 1, 4]
B = [
       [2, 3, 2],
       [1, 4, 5],
       [4, 4, 1]
     ]
def solve(A, B):
    n = len(A)
    print(A)
    M=[0]*n
    print(M)
    print("###")
    for l,r,c in B:
       M[l-1]+=c
       if r<=n-1:
           M[r]-=c
       print(M)
    M[0] = A[0]+M[0]
    print(M)
    for i in range(1,n):
        M[i]+=M[i-1]+A[i]-A[i-1]


    return M

solve(A,B)

[1, 2, 1, 4]
[0, 0, 0, 0]
###
[0, 2, 0, -2]
[5, 2, 0, -2]
[5, 2, 0, -1]
[6, 2, 0, -1]


[6, 9, 8, 10]

Chocolate Distribution
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
Given an array A of N integers where the i-th element represent the number of chocolates in the i-th packet.

There are B number of students, the task is to distribute chocolate packets following below conditions:

1. Each student gets one packets.
2. The difference between the number of chocolates given to any two students is minimum.
Return the minimum difference (that can be achieved) between the student who gets minimum number of chocolates and the student who gets maximum number of chocolates.

Note: If you can't give each student 1 packet, return 0.



Problem Constraints
0 <= N <= 10^5
1 <= A[i] <= 10^7
0 <= B <= 10^5


Input Format
The first argument contains an integer array A.

The second argument contains an integer B.



Output Format
Return the minimum difference (that can be achieved) between maximum and minimum number of chocolates distributed.


Example Input
Input:

  A : [3, 4, 1, 9, 56, 7, 9, 12]
  B : 5


Example Output
Output:

  6


Example Explanation
Explanation:

  
  We can choose the packets with chocolates = [3, 4, 9, 7, 9]
  The difference between maximum and minimum is 9-3 = 6

In [137]:
A = [3, 4, 1, 9, 56, 7, 9, 12]
#A = [1,3,4,7,9,9,12,56]

B = 5

def chocolate_distribution(A,B):
    n=len(A)
    if n < B or B==0 :
        return 0
    A.sort()
    diff = A[B-1] - A[0]
    if n == B:
        return diff
    start = 1
    for i in range(B,n):
        if A[i] - A[start] < diff:
            diff = A[i] - A[start]
        start+=1

    return diff
    
chocolate_distribution(A,B)
            


6

Rain Water Trapped
Unsolved
feature icon
Get your doubts resolved blazing fast with Chat GPT Help
Check Chat GPT
feature icon
Using hints is now penalty free
Use Hint
Problem Description
Given a vector A of non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.



Problem Constraints
1 <= |A| <= 100000



Input Format
First and only argument is the vector A



Output Format
Return one integer, the answer to the question



Example Input
Input 1:

A = [0, 1, 0, 2]
Input 2:

A = [1, 2]


Example Output
Output 1:

1
Output 2:

0


Example Explanation
Explanation 1:

1 unit is trapped on top of the 3rd element.
Explanation 2:

No water is trapped.

In [167]:
"""
1. if single element then no water 
2. if two elements then no water
3. Three elements . first and last index can't contain water . 
"""
A=[1,2,0,1,2,1,5,1,2,1,4,1,5,1,0,0,0,1,0,3]
def trap_water(A):
    n = len(A)
    height=0
    for i in range(1,n-1):
        left_max = max(A[:i:1])
        right_max = max(A[i+1:n:])
        if A[i] < left_max and A[i] < right_max:
            height += (min(left_max,right_max) - A[i])
            print((i,height),end=" ")
    print()
    print(height)

def trap_water_1(Array_Buildings_Height): # verbose
    no_of_buildings = len(Array_Buildings_Height)
    initial_water_unit = 0
    total_water_unit = initial_water_unit
    for index,height in enumerate(Array_Buildings_Height):
        #print(index,height)
        if index == 0 or index == no_of_buildings -1 :
            left_max = 0
            right_max = 0
        else:
            left_max = max(Array_Buildings_Height[:index:1])
            right_max = max(Array_Buildings_Height[index+1:no_of_buildings:1])
        if height < left_max and height < right_max:
            total_water_unit += min(left_max,right_max) - height
    return total_water_unit

def trap_water_2(A):
    n = len(A)
    water = 0
    for i in range(1,n-1):
        left_max = 0
        for j in range(i):
            if A[j] > left_max:
                left_max = A[j]
        right_max = 0
        for k in range(i+1,n):
            if A[k] > right_max:
                right_max = A[k]
        level = 0
        if A[i] < left_max and A[i] < right_max:
            level = min(left_max,right_max) - A[i]
        water+=level

    return water

def prefix_sum(A):
    n = len(A)
    P=[]
    sum_p =0 
    for i in range(n):
        sum_p += A[i]
        P.append(sum_p)
    return P

def prefix_max(A):
    n = len(A)
    P = []
    P.append(A[0])
    for i in range(1,n):
        P.append(max(P[i-1],A[i]))
    return P

def suffix_sum(A):
    n = len(A)
    S = []
    sum_s = 0
    # 0 to n-1.n-1 to 0
    for i in reversed(range(n)):
        #print(i,A[i])
        sum_s+=A[i]
        S.append(sum_s)
    return S[::-1]

def suffix_max(A):
    n = len(A)
    S = A[:] # cant be used for 2d array 
    print(S)
    for i in reversed(range(n-1)): # think 0 to n-2 and then reverse n-2 to 0 . 
        S[i] = max(S[i+1],A[i])

    return S

def trap_water_3(A):#optimized using prefix max using carry forward and no need to store and suffix max on the right side by storing it 
    n = len(A)
    S = suffix_max(A)
    total_water = 0
    left_max = 0 
    for i in range(n):
        left_max = max(left_max,A[i])
        right_max = S[i]
        total_water += min(left_max,right_max) - A[i]
    return total_water






        
#trap_water(A)
#trap_water_1(A)
#trap_water_2(A)
#prefix_sum(A)
#prefix_max(A)
#suffix_sum(A)
#suffix_max(A)
#trap_water_3(A)
S = A[:]
print(S)
type(S)

[1, 2, 0, 1, 2, 1, 5, 1, 2, 1, 4, 1, 5, 1, 0, 0, 0, 1, 0, 3]


list